pkg/sysregistriesv2/system_registries_v2.go

```diff
@@ -189,6 +196,235 @@ func getV1Registries(config *tomlConfig) ([]Registry, error) {
+func registryIsMatch(reg Registry) bool {
+	return reg.Insecure || reg.Blocked
+}
+
+// registryMatches returns true if the candidate location matches the pattern.
+// Recognized patterns include:
+// * IPv4[/class][:port][/path]
+// * IPv6[/class][:port][/path]
+// * hostname[:port][/path]
+// where IPv4, IPv6, and hostname may include wildcard characters.  The path
+// is implicitly terminated with "/", and individual components may also
+// include wildcard characters.
+func registryMatches(pattern, candidate string) (bool, error) {
+	defaultRegistryPort := "5000"
+	phost, pport, ppath, err := parseMatchPattern(pattern)
+	if err != nil {
+		return false, err
+	}
+	chost, cport, cpath, err := parseMatchCandidate(candidate)
+	if err != nil {
+		return false, err
+	}
+	// Check the host component
+	_, pipnet, perr := net.ParseCIDR(phost)
+	cip := net.ParseIP(chost)
+	if cip == nil && perr == nil {
+		// can't match a non-IP against a CIDR address
+		return false, nil
+	}
+	if perr == nil {
+		// pattern is CIDR and the candidate is an IP
+		if !pipnet.Contains(cip) {
+			return false, nil
+		}
+	} else {
+		// pattern and candidate are both names
+		hosts, err := path.Match(phost, chost)
+		if err != nil {
+			return false, fmt.Errorf("error matching candidate host %q against pattern host %q: %v", chost, phost, err)
+		}
+		if !hosts {
+			return false, nil
+		}
+	}
+	// Check the ports, if specified, or default ports
+	if pport == "" {
+		pport = defaultRegistryPort
+	}
+	if cport == "" {
+		cport = defaultRegistryPort
+	}
+	if pport != "*" && pport != cport {
+		return false, nil
+	}
+	// Compare paths, only so far as the pattern supplies one
+	pdir, ppath := path.Split(ppath)
+	cdir, cpath := path.Split(cpath)
+	for pdir != "" {
+		if cdir == "" {
+			return false, nil
+		}
+		matches, err := path.Match(pdir, cdir)
+		if err != nil {
+			return false, fmt.Errorf("error comparing path component %q to pattern component %q: %v", cdir, pdir, err)
+		}
+		if !matches {
+			return false, nil
+		}
+		pdir, ppath = path.Split(ppath)
+		cdir, cpath = path.Split(cpath)
+	}
+	return true, nil
+}
+
+// isBogusIP checks if the passed-in value looks superficially like an IPv4
+// address, but contains an invalid octet value.
+func isBogusIP(candidate string) bool {
+	components := strings.Split(candidate, ".")
+	// no problem if there aren't 4 .-separated components
+	if len(components) != 4 {
+		return false
+	}
+	for _, i := range components {
+		// or if any of them don't look like numbers
+		if len(strings.FieldsFunc(i, func(r rune) bool { return unicode.IsDigit(r) })) != 0 {
+			return false
+		}
+	}
+	for _, i := range components {
+		// but if they do, and they're invalid values, that's a problem
+		if _, err := strconv.ParseUint(i, 10, 8); err != nil {
+			return true
+		}
+	}
+	return false
+}
+
+// parseMatchPattern splits up matching information for registries.
+func parseMatchPattern(pattern string) (host, port, path string, err error) {
```

Comment:
(These parsers are simultaneously a lot of code to maintain / modify, and very non-obvious about allowing/rejecting corner cases and invalid inputs.)